nutils/sample.py

```diff
@@ -85,16 +85,55 @@ def __init__(self, transforms:types.tuple[transformseq.stricttransforms], points
     assert len(points) == len(index)
     assert len(transforms) >= 1
     assert all(len(t) == len(points) for t in transforms)
+    assert all(t.roots == transforms[0].roots for t in transforms[1:])
     self.nelems = len(transforms[0])
     self.transforms = transforms
     self.points = points
     self.index = index
     self.npoints = sum(p.npoints for p in points)
     self.ndims = transforms[0].fromdims
+    self.roots = self.transforms[0].roots
 
   def __repr__(self):
     return '{}<{}D, {} elems, {} points>'.format(type(self).__qualname__, self.ndims, self.nelems, self.npoints)
 
+  def _common_eval(self, ielem):
+    points = self.points[ielem]
+    transforms = tuple(t[ielem] for t in self.transforms)
+    detJ = 1
+    coords = []
+    tangents = []
+    normals = []
+    globalndims = 0
+    for root, trans in zip(self.roots, transforms[0]):
+      ndims = trans[-1].fromdims if trans else root.ndims
+      coords.append(points.coords[:,globalndims:globalndims+ndims])
+      globalndims += ndims
+      if trans and trans[-1].fromdims == 0 or root.ndims == 0:
+        T = numpy.zeros((root.ndims,0))
+        N = numpy.eye(root.ndims)
+      elif not trans:
+        T = numpy.eye(root.ndims)
+        N = numpy.zeros((root.ndims,0))
+      else:
+        V = transform.linearfrom(trans, root.ndims)
+        J = V[:,:trans[-1].fromdims]
+        if J.shape[1] == 1:
+          detJ *= abs(numpy.dot(J.T, J)[0,0])**0.5
+        else:
+          detJ *= abs(numpy.linalg.det(numpy.dot(J.T, J)))**0.5
+
+        V = numpy.array(V, copy=True, dtype=float)
+        for i in range(root.ndims):
+          if i > 0:
+            V[:,i] -= numpy.dot(V[:,:i], numpy.dot(V[:,:i].T, V[:,i]))
+          V[:,i] /= numpy.linalg.norm(V[:,i])
+        T, N = V[:,:trans[-1].fromdims], V[:,trans[-1].fromdims:]
```

Comment:
shouldn't there be an additional case here for trans[-1].fromdims == root.ndims so that we can take the direct determinant of linear?